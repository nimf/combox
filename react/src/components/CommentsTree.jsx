import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Comment } from 'semantic-ui-react';
import CommentNode from './CommentNode';

const propTypes = {
  comments: PropTypes.object.isRequired,
  currentCommentId: PropTypes.number,
  onFocusedRendered: PropTypes.func.isRequired,
};
const defaultProps = {
  currentCommentId: null,
};

class CommentsTree extends Component {
  componentDidUpdate() {
    if (this.props.comments.focusedId !== null &&
        this.props.comments.byId[this.props.comments.focusedId].parentId ===
        this.props.currentCommentId) {
      // If we rendered a comment that must receive focus, report it
      this.props.onFocusedRendered();
    }
  }
  render() {
    const currentComments = this.props.comments.allIds.map(id => this.props.comments.byId[id])
      .filter(c => c.parentId === this.props.currentCommentId);
    if (currentComments.length === 0) return null;
    const listOfComments = currentComments.map(comment =>
      (
        <CommentNode
          key={comment.id}
          id={comment.id}
          authorName={comment.authorName}
          isGuest={comment.isGuest}
          createdAt={comment.createdAt}
          message={comment.message}
          fresh={this.props.comments.focusedId === comment.id}
        >
          <CommentsTree
            comments={this.props.comments}
            currentCommentId={comment.id}
            onFocusedRendered={this.props.onFocusedRendered}
          />
        </CommentNode>
      ));
    return (
      <Comment.Group threaded={this.props.currentCommentId === null}>
        {listOfComments}
      </Comment.Group>
    );
  }
}

CommentsTree.propTypes = propTypes;
CommentsTree.defaultProps = defaultProps;

export default CommentsTree;
