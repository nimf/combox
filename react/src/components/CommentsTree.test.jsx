import React from 'react';
import renderer from 'react-test-renderer';
import CommentsTree from './CommentsTree';
import commentsFixture from '../fixtures/comments';

describe('CommentsTree renders', () => {
  test('renders root comments and below', () => {
    const tree = renderer.create(<CommentsTree
      comments={commentsFixture}
      onFocusedRendered={() => {}}
      onOpenNewComments={() => {}}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
