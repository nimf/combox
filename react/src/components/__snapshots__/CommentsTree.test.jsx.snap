// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsTree renders renders root comments and below 1`] = `
<div
  className="ui threaded comments"
>
  <div
    className="comment"
    id="comment-1"
  >
    <a
      className="avatar"
    >
      <img
        src="https://www.gravatar.com/avatar/00000000000000000000000000000000?s=100"
      />
    </a>
    <div
      className="content"
    >
      <span
        className="author"
      >
        <span>
          Anonymous
        </span>
        <div
          className="ui mini label"
          onClick={[Function]}
        >
          Guest
        </div>
      </span>
      <div
        className="metadata"
      >
        <div
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          3 days ago
        </div>
      </div>
      <div
        className="text"
      >
        <p>
          Test comment 1
        </p>
      </div>
      <div
        className="actions"
      >
        Reply
      </div>
    </div>
    <div
      className="ui comments"
    >
      <div
        className="comment"
        id="comment-2"
      >
        <a
          className="avatar"
        >
          <img
            src="https://www.gravatar.com/avatar/00000000000000000000000000000000?s=100"
          />
        </a>
        <div
          className="content"
        >
          <span
            className="author"
          >
            <span>
              Anonymous
            </span>
            <div
              className="ui mini label"
              onClick={[Function]}
            >
              Guest
            </div>
          </span>
          <div
            className="metadata"
          >
            <div
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              2 days ago
            </div>
          </div>
          <div
            className="text"
          >
            <p>
              Test comment 2
            </p>
          </div>
          <div
            className="actions"
          >
            Reply
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="comment"
    id="comment-3"
  >
    <a
      className="avatar"
    >
      <img
        src="https://www.gravatar.com/avatar/00000000000000000000000000000000?s=100"
      />
    </a>
    <div
      className="content"
    >
      <span
        className="author"
      >
        <span>
          Anonymous
        </span>
        <div
          className="ui mini label"
          onClick={[Function]}
        >
          Guest
        </div>
      </span>
      <div
        className="metadata"
      >
        <div
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          a day ago
        </div>
      </div>
      <div
        className="text"
      >
        <p>
          Test comment 3
        </p>
      </div>
      <div
        className="actions"
      >
        Reply
      </div>
    </div>
  </div>
</div>
`;
